# Customer churn prediction
This notebook uses a subset (128MB) of the full dataset available (12GB) for the convinence of manipulating data and build machine learning model locally. 

In [233]:
# import libraries
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, sum, count, udf, to_timestamp, min, max, avg, when, length
from pyspark.sql.types import StringType
from pyspark.sql.types import IntegerType


import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline


In [234]:
# create a Spark session
spark = SparkSession \
    .builder \
    .appName("Sparkify") \
    .getOrCreate()

# Load and Clean Dataset
In this workspace, the mini-dataset file is `mini_sparkify_event_data.json`. Load and clean the dataset, checking for invalid or missing data - for example, records without userids or sessionids. 

### 1. Load and overview of the data

In [235]:
# Load and read data as dataframe
path = "mini_sparkify_event_data.json"
df = spark.read.json(path)
df.printSchema()

root
 |-- artist: string (nullable = true)
 |-- auth: string (nullable = true)
 |-- firstName: string (nullable = true)
 |-- gender: string (nullable = true)
 |-- itemInSession: long (nullable = true)
 |-- lastName: string (nullable = true)
 |-- length: double (nullable = true)
 |-- level: string (nullable = true)
 |-- location: string (nullable = true)
 |-- method: string (nullable = true)
 |-- page: string (nullable = true)
 |-- registration: long (nullable = true)
 |-- sessionId: long (nullable = true)
 |-- song: string (nullable = true)
 |-- status: long (nullable = true)
 |-- ts: long (nullable = true)
 |-- userAgent: string (nullable = true)
 |-- userId: string (nullable = true)



In [255]:
num_rows = df.count()
num_cols = len(df.columns)
print('The data set contains {} rows and {} columns'.format(num_rows, num_cols))

The data set contains 286500 rows and 18 columns


In [237]:
# Show the fist row of the dataframe
df.show(n=1)

+--------------+---------+---------+------+-------------+--------+---------+-----+---------------+------+--------+-------------+---------+---------+------+-------------+--------------------+------+
|        artist|     auth|firstName|gender|itemInSession|lastName|   length|level|       location|method|    page| registration|sessionId|     song|status|           ts|           userAgent|userId|
+--------------+---------+---------+------+-------------+--------+---------+-----+---------------+------+--------+-------------+---------+---------+------+-------------+--------------------+------+
|Martha Tilston|Logged In|    Colin|     M|           50| Freeman|277.89016| paid|Bakersfield, CA|   PUT|NextSong|1538173362000|       29|Rockpools|   200|1538352117000|Mozilla/5.0 (Wind...|    30|
+--------------+---------+---------+------+-------------+--------+---------+-----+---------------+------+--------+-------------+---------+---------+------+-------------+--------------------+------+
only showi

In [238]:
# Show all the columns
df.columns

['artist',
 'auth',
 'firstName',
 'gender',
 'itemInSession',
 'lastName',
 'length',
 'level',
 'location',
 'method',
 'page',
 'registration',
 'sessionId',
 'song',
 'status',
 'ts',
 'userAgent',
 'userId']

### 2. Check missing values and clean data

In [239]:
def check_missing_values(df):
    """Check the number of missing values in each columns and calculate the missing rate
    
    Arguments:
        df {dataframe} -- spark dataframe
        
    Return: None    
    """
    print('Column name         null count   null rate\n')
    for c in df.columns:
        null_count = df.filter(df[c].isNull() | (df[c] == '') ).count() # Check whether each column has null value or empty string
        null_rate = null_count/df.count() 
        print('{:15} | {:10} | {:10}'.format(c, null_count, null_rate))

In [240]:
check_missing_values(df)

Column name         null count   null rate

artist          |      58392 | 0.20381151832460734
auth            |          0 |        0.0
firstName       |       8346 | 0.029130890052356022
gender          |       8346 | 0.029130890052356022
itemInSession   |          0 |        0.0
lastName        |       8346 | 0.029130890052356022
length          |      58392 | 0.20381151832460734
level           |          0 |        0.0
location        |       8346 | 0.029130890052356022
method          |          0 |        0.0
page            |          0 |        0.0
registration    |       8346 | 0.029130890052356022
sessionId       |          0 |        0.0
song            |      58392 | 0.20381151832460734
status          |          0 |        0.0
ts              |          0 |        0.0
userAgent       |       8346 | 0.029130890052356022
userId          |       8346 | 0.029130890052356022


It is shown above that columns such as 'firstName', 'gender', 'lastName', 'location', 'registration' and 'userAgent' all have 8346 missing values as 'userId', because these information are all 'userId' related. Since our purpose in this project is to analyze user behavior, it would be safe to drop those rows that have no user information. 


In [296]:
# Drop rows with missing user id
df_dropId = df.filter((df['userId'] != '' ) )

In [284]:
# Total rows left
df_dropId.count()

278154

Another case is that 3 song-related columns 'artist', 'length' and 'song' have the same number of missing values. Let's check their interaction with 'page' column to get some idea. 

In [287]:
df.groupby('page').agg({'artist':'count', 'length':'count', 'song':'count'}).show(25)

+--------------------+-----------+-------------+-------------+
|                page|count(song)|count(artist)|count(length)|
+--------------------+-----------+-------------+-------------+
|              Cancel|          0|            0|            0|
|    Submit Downgrade|          0|            0|            0|
|         Thumbs Down|          0|            0|            0|
|                Home|          0|            0|            0|
|           Downgrade|          0|            0|            0|
|         Roll Advert|          0|            0|            0|
|              Logout|          0|            0|            0|
|       Save Settings|          0|            0|            0|
|Cancellation Conf...|          0|            0|            0|
|               About|          0|            0|            0|
| Submit Registration|          0|            0|            0|
|            Settings|          0|            0|            0|
|               Login|          0|            0|       

The three song related columns exclusively relate with 'NextSong' in page column. A user is unlikely to churn simply because of the names of the artist or song. So these columns could be dropped. As for the lengh of the song, We will explore it further in feature engineering section.

In [ ]:
# Drop 'artist' and 'song' columns

In [289]:
df_drop = df_dropId.drop('artist', 'song')

Now, with missing user ids removed, let's see how many users in total are registered. 

In [290]:
# Check number of unique users
n_users = df.select('userId').dropDuplicates().count()
print('There are {} unique users in the dataset.'.format(n_users))

There are 226 unique users in the dataset.


Since user ids are unique values to idendify users, we don't need extra 'firstName' and 'lastName' columns. Additionally, a user is unlikely to churn simply because of his/her own location. So we will drop these columns, too.

In [291]:
df_drop = df_drop.drop('firstName','lastName', 'location')

In [292]:
df_drop.show(5)

+---------+------+-------------+---------+-----+------+--------+-------------+---------+------+-------------+--------------------+------+
|     auth|gender|itemInSession|   length|level|method|    page| registration|sessionId|status|           ts|           userAgent|userId|
+---------+------+-------------+---------+-----+------+--------+-------------+---------+------+-------------+--------------------+------+
|Logged In|     M|           50|277.89016| paid|   PUT|NextSong|1538173362000|       29|   200|1538352117000|Mozilla/5.0 (Wind...|    30|
|Logged In|     M|           79|236.09424| free|   PUT|NextSong|1538331630000|        8|   200|1538352180000|"Mozilla/5.0 (Win...|     9|
|Logged In|     M|           51| 282.8273| paid|   PUT|NextSong|1538173362000|       29|   200|1538352394000|Mozilla/5.0 (Wind...|    30|
|Logged In|     M|           80|262.71302| free|   PUT|NextSong|1538331630000|        8|   200|1538352416000|"Mozilla/5.0 (Win...|     9|
|Logged In|     M|           52|22

In [297]:
# Take a look at numerical columns
df_drop.select('itemInSession','registration','length','sessionId', 'status','ts').describe().show()


+-------+------------------+--------------------+------------------+------------------+------------------+--------------------+
|summary|     itemInSession|        registration|            length|         sessionId|            status|                  ts|
+-------+------------------+--------------------+------------------+------------------+------------------+--------------------+
|  count|            278154|              278154|            228108|            278154|            278154|              278154|
|   mean|114.89918174824018|1.535358834085619...|249.11718197783583|1042.5616241362698|209.10321620397335|1.540958915431790...|
| stddev|129.85172939948959|3.2913216163281236E9| 99.23517921058313| 726.5010362219864|  30.1513888513279|1.5068287123356583E9|
|    min|                 0|       1521380675000|           0.78322|                 1|               200|       1538352117000|
|    max|              1321|       1543247354000|        3024.66567|              2474|               40

In [298]:
# Take a look at categorical columns
# and count occurences in each category in each column
cat_cols = ['auth', 'gender', 'level', 'method', 'page', 'userAgent', 'userId']
for c in cat_cols:
    grouped = df_drop.groupby(c).count().show()

+---------+------+
|     auth| count|
+---------+------+
|Cancelled|    52|
|Logged In|278102|
+---------+------+

+------+------+
|gender| count|
+------+------+
|     F|154578|
|     M|123576|
+------+------+

+-----+------+
|level| count|
+-----+------+
| free| 55721|
| paid|222433|
+-----+------+

+------+------+
|method| count|
+------+------+
|   PUT|257818|
|   GET| 20336|
+------+------+

+--------------------+------+
|                page| count|
+--------------------+------+
|              Cancel|    52|
|    Submit Downgrade|    63|
|         Thumbs Down|  2546|
|                Home| 10082|
|           Downgrade|  2055|
|         Roll Advert|  3933|
|              Logout|  3226|
|       Save Settings|   310|
|Cancellation Conf...|    52|
|               About|   495|
|            Settings|  1514|
|     Add to Playlist|  6526|
|          Add Friend|  4277|
|            NextSong|228108|
|           Thumbs Up| 12551|
|                Help|  1454|
|             Upgrade|   499|


The 'status', 'method' and 'userAgent'column contains network related records, can also be dropped. 

In [ ]:
df_drop = df_drop.drop('status','method','userAgent')

In [304]:
# Show cleaned dataset
df_drop.show(5)

+---------+------+-------------+---------+-----+--------+-------------+---------+-------------+------+
|     auth|gender|itemInSession|   length|level|    page| registration|sessionId|           ts|userId|
+---------+------+-------------+---------+-----+--------+-------------+---------+-------------+------+
|Logged In|     M|           50|277.89016| paid|NextSong|1538173362000|       29|1538352117000|    30|
|Logged In|     M|           79|236.09424| free|NextSong|1538331630000|        8|1538352180000|     9|
|Logged In|     M|           51| 282.8273| paid|NextSong|1538173362000|       29|1538352394000|    30|
|Logged In|     M|           80|262.71302| free|NextSong|1538331630000|        8|1538352416000|     9|
|Logged In|     M|           52|223.60771| paid|NextSong|1538173362000|       29|1538352676000|    30|
+---------+------+-------------+---------+-----+--------+-------------+---------+-------------+------+
only showing top 5 rows



Above we cleaned the dataframe by dropping  rows with mising user id and columns that are not relevant. The next step is to define churn and perform exploratory data analysis.  

In [301]:
# Count occurrences in each page category
df_drop.groupby('page').count().show(30)

+--------------------+------+
|                page| count|
+--------------------+------+
|              Cancel|    52|
|    Submit Downgrade|    63|
|         Thumbs Down|  2546|
|                Home| 10082|
|           Downgrade|  2055|
|         Roll Advert|  3933|
|              Logout|  3226|
|       Save Settings|   310|
|Cancellation Conf...|    52|
|               About|   495|
|            Settings|  1514|
|     Add to Playlist|  6526|
|          Add Friend|  4277|
|            NextSong|228108|
|           Thumbs Up| 12551|
|                Help|  1454|
|             Upgrade|   499|
|               Error|   252|
|      Submit Upgrade|   159|
+--------------------+------+



# Exploratory Data Analysis

### Define Churn

Once you've done some preliminary analysis, create a column `Churn` to use as the label for your model. I suggest using the `Cancellation Confirmation` events to define your churn, which happen for both paid and free users. As a bonus task, you can also look into the `Downgrade` events.

1. At the moment, the subscription ends and renewal doesn’t happen, or

2. At the moment of the cancelation.

3. Downgrade

### Explore Data
Once you've defined churn, perform some exploratory data analysis to observe the behavior for users who stayed vs users who churned. You can start by exploring aggregates on these two groups of users, observing how much of a specific action they experienced per a certain time unit or number of songs played.

# Feature Engineering
Once you've familiarized yourself with the data, build out the features you find promising to train your model on. To work with the full dataset, you can follow the following steps.
- Write a script to extract the necessary features from the smaller subset of data
- Ensure that your script is scalable, using the best practices discussed in Lesson 3
- Try your script on the full data set, debugging your script if necessary

If you are working in the classroom workspace, you can just extract features based on the small subset of data contained here. Be sure to transfer over this work to the larger dataset when you work on your Spark cluster.

# Modeling
Split the full dataset into train, test, and validation sets. Test out several of the machine learning methods you learned. Evaluate the accuracy of the various models, tuning parameters as necessary. Determine your winning model based on test accuracy and report results on the validation set. Since the churned users are a fairly small subset, I suggest using F1 score as the metric to optimize.

# Final Steps
Clean up your code, adding comments and renaming variables to make the code easier to read and maintain. Refer to the Spark Project Overview page and Data Scientist Capstone Project Rubric to make sure you are including all components of the capstone project and meet all expectations. Remember, this includes thorough documentation in a README file in a Github repository, as well as a web app or blog post.